File: generators/reference/kernels/norm_layer.py
"""norm.layer@1.0.0 — (x − mean) · rsqrt(var + eps) · weight + bias over `width`, the variance
biased, as torch's LayerNorm and BERT compute it.

| branch / record | status      |
|-----------------|-------------|
| eps, weight, bias | implemented |
"""
import torch.nn.functional as F
from kernels._common import refuse_unknown, supports_from, w

CONTRACT = ("norm.layer", "1.0.0")


CAPABILITIES = {"arguments": {"width": "any", "eps": "any"}, "states": []}


def supports(arguments):
    return supports_from(CAPABILITIES, arguments)

def run(ctx, arguments, inputs, params, states, physical=None):
    x = inputs['input']
    return {'output': F.layer_norm(x, (x.shape[-1],), w(ctx, params['weight']), w(ctx, params['bias']), arguments['eps'])}

File: generators/zml/primitives/norm_layer.zig
//! norm.layer@1.0.0 — (x − mean) · rsqrt(var + eps) · weight + bias over `width`.
//!
//! The variance is biased, as torch's LayerNorm and BERT compute it — which is what
//! `zml.nn.normalizeVariance` does too.

const zml = @import("zml");

const p = @import("../primitive.zig");

pub const primitive: p.Primitive = .{
    .name = "norm.layer",
    .version = "1.0.0",
    .run = run,
    .capabilities =
    \\{"arguments": {"width": "any", "eps": "any"}, "states": []}
    ,
};

fn run(ctx: *p.Ctx, call: p.Call) ![]const p.Binding {
    const eps: f32 = @floatCast(try call.requireFloat("eps"));
    const x = p.features(call.inputs.must("input"));

    const norm: zml.nn.LayerNorm = .{
        .weight = p.features(call.params.must("weight")).convert(x.dtype()),
        .bias = p.features(call.params.must("bias")).convert(x.dtype()),
        .eps = eps,
    };
    return p.one(ctx, "output", norm.forward(x));
}
